forem/forem
# API V1 transition

## What type of PR is this? (check all applicable)


## Description

This is a first step (leap?) at upgrading our `V0` API to a `V1` version with a transition period where we support both versions. The transition period will likely end when we completely sunset our V0 and remove the FeatureFlag of `V1`.

The API controller logic from files in `/app/controllers/api/v0/*.rb` was extracted to corresponding concerns (`/app/controllers/concerns/api/*_controller.rb`) so it can be reused in both v0 and v1 (the logic remains the same for both of them).

The difference between versions are the `before_action` hooks (these continue to live in `/app/controllers/api/v0/*.rb` and `/app/controllers/api/v1/*.rb`) where we are enabling new authentication restrictions.

**Important notes:**
- Specs for the API `V0` remain unchanged (V0 should continue to work as expected)
- This PR includes a routing constraint on `V1` endpoints that ensures `V1` is only accessible with `:api_v1` FeatureFlag

## Related Tickets & Documents



- Related Issue #17668 
- Related Issue #17657

## QA Instructions, Screenshots, Recordings

Take a look at the existing specs. Some suggested/useful commands to test subsets of these specs locally are:
- `bundle exec spec/requests/api/v0`
   - Run all API `V0` specs
- `bundle exec spec/requests/api/v1`
   - Run all API `V1` specs
- `bundle exec spec/requests/api`
   - Run all API specs (`V0` + `V1`)

You can also checkout the app locally and register a new account + go through onboarding. That is one of the few critical areas that API migration touches, but given that (1) the new API is behind a Feature Flag, and (2) `V0` remains untouched (and is the default) we shouldn't see any effect on the app itself.

### UI accessibility concerns?

None - backend update only

## Added/updated tests?


## [Forem core team only] How will this change be communicated?


## [optional] Are there any post deployment tasks we need to perform?

Merge this PR only while SRE/ops team is online and able to help out monitor in case any unexpected effect makes it way to production

## [optional] What gif best describes this PR or how it makes you feel?

![hello, computer?](https://media.giphy.com/media/PxSFAnuubLkSA/giphy.gif)
